core/src/HIP/Kokkos_HIP_ReduceScan.hpp

```diff
@@ -49,8 +49,326 @@
 
 #if defined(__HIPCC__)
 
+#include <HIP/Kokkos_HIP_Vectorization.hpp>
+
 namespace Kokkos {
 namespace Impl {
+
+/* Algorithmic constraints:
+ *   (a) threads with the same hipThreadIdx_x have same value
+ *   (b) hipBlockDim_x == power of two
+ *   (x) hipBlockDim_z == 1
+ */
+template <typename ValueType, typename JoinOp,
+          typename std::enable_if<!Kokkos::is_reducer<ValueType>::value,
+                                  int>::type = 0>
+__device__ inline void hip_intra_warp_reduction(
+    ValueType& result, JoinOp const& join,
+    uint32_t const max_active_thread = hipBlockDim_y) {
+  unsigned int shift = 1;
+
+  // Reduce over values from threads with different hipThreadIdx_y
+  unsigned int constexpr warp_size =
+      Kokkos::Experimental::Impl::HIPTraits::WarpSize;
+  while (hipBlockDim_x * shift < warp_size) {
+    ValueType const tmp = Kokkos::Experimental::shfl_down(
+        result, hipBlockDim_x * shift, warp_size);
+    // Only join the if upper thread is active (this allows non power of two for
+    // hipBlockDim_y)
+    if (hipThreadIdx_y + shift < max_active_thread) join(result, tmp);
+    shift *= 2;
+  }
+
+  // Broadcast the result to all the threads in the warp
+  result = Kokkos::Experimental::shfl(result, 0, warp_size);
+}
+
+template <typename ValueType, typename JoinOp,
+          typename std::enable_if<!Kokkos::is_reducer<ValueType>::value,
+                                  int>::type = 0>
+__device__ inline void hip_inter_warp_reduction(
+    ValueType& value, const JoinOp& join,
+    const int max_active_thread = hipBlockDim_y) {
+  unsigned int constexpr warp_size =
+      Kokkos::Experimental::Impl::HIPTraits::WarpSize;
+  int constexpr step_width = 8;
+  // Depending on the ValueType __shared__ memory must be aligned up to 8 byte
+  // boundaries. The reason not to use ValueType directly is that for types with
+  // constructors it could lead to race conditions.
+  __shared__ double sh_result[(sizeof(ValueType) + 7) / 8 * step_width];
+  ValueType* result = reinterpret_cast<ValueType*>(&sh_result);
+  int const step    = warp_size / hipBlockDim_x;
+  int shift         = step_width;
+  int const id = hipThreadIdx_y % step == 0 ? hipThreadIdx_y / step : 65000;
```

Comment:
Is 65000 just a really large value so we don't do anything below? If so, let's just choose `INT_MAX` or the like and add a comment.